### PR TITLE
Don't call _emscripten_futex_wake if not a function

### LINKED ISSUE #8446: When a pthread fails initialisation the wrong error is thrown
Unless I'm missing something, if `Module['dynCall_ii'](e.data.start_routine, e.data.arg);` ([here](https://github.com/emscripten-core/emscripten/blob/a51cda8f97accdb42449e3e810a702d8007e2512/src/worker.js#L191)) throws, `Module['_emscripten_futex_wake']` ([here](https://github.com/emscripten-core/emscripten/blob/a51cda8f97accdb42449e3e810a702d8007e2512/src/worker.js#L206)) will be undefined. This means that whatever error was actually caught will never be re-thrown, and `Module._emscripten_futex_wake is not a function` will always be thrown instead.

In ideal situations this should of course never occur, but it does lead to some misleading errors during development/debugging.

I'm tempted to submit a PR just removing that `_emscripten_futex_wake`, because I'm not sure what value it adds. If this function call actually has a use I could just open a PR when the called is wrapped in a `if (typeof(Module["_emscripten_futex_wake"]) === function) {...}`

Thoughts?

### PR DESCRIPTION
**Pull request by VirtualTim**
This can happen if the thread throws on initialisation. Trying to call a function that doesn't exist will throw, hiding the previously caught exception.
Fixes #8446.

### CODE REVIEW

**Comment by kripken** [comment]
How does this hide the previous exception? Both should be logged, I'd think?

I tend to think that we should not add code that tries to make things work if the thread failed to even initialize - then the entire state of the world is just bad, and nothing can be expected to work. There are any number of other possible problems in that case, I'd guess.

**Comment by VirtualTim** [comment]
So the code I modified is in a try-catch. The logic is basically:
If a thread throws, 
1. check if the exception is not "Canceled! | SimulateInfiniteLoop | pthread_exit"
2. mark the thread as stopped, 
3. restart threads waiting on this one, 
4. then rethrow the caught error.

If restarting threads (3.) throws, then the previously caught error never gets re-thrown (4.).

The point of this isn't to try and keep things working, its to try and make sure that the error that gets bubbled up is the initial error, not `Module._emscripten_futex_wake is not a function`. The purpose of this is for developers, not for the end user.

Basically this is about making the thrown error the correct/original one.

**Comment by kripken** [comment]
I see. Ok, makes sense to improve the error for developers, I agree. How about doing something like this, though?
```
#if ASSERTIONS
  assert(..that thing is a function.., ".. error message that explains pthreads were not set up properly etc. " + original error message);
#endif
```
Putting it behind ASSERTIONS ensures it doesn't add overhead in general, and I think it's clearer as an assert (rather than an if that just skips something).

**Comment by VirtualTim** [comment]
Yeah I guess I can do that. I just thought overhead didn't really matter since it only reaches this check if something goes fatally wrong.

**Comment by kripken** [comment]
The overhead is minor, I agree, but I think it's also a matter of clarity - an assertion directly says what we expect to be true there.

**Comment by VirtualTim** [comment]
I guess, but you lose useful information like the stack trace. Instead you get the stack trace of the wrong exception. And in builds without ASSERTIONS defined the wrong error is still thrown.
I guess I could put an `#if ASSERTIONS` around that line? Would that be acceptable?

**Comment by sbc100** [comment]
Whichever way you do it can you make you there is comment about why `_emscripten_futex_wake` might end up not being a function at this point in the code?

File: src/worker.js [hunk 1]
```diff
@@ -203,7 +203,9 @@ this.onmessage = function(e) {
         } else {
           Atomics.store(HEAPU32, (threadInfoStruct + 4 /*C_STRUCTS.pthread.threadExitCode*/ ) >> 2, (e instanceof {{{ makeAsmGlobalAccessInPthread('ExitStatus') }}}) ? e.status : -2 /*A custom entry specific to Emscripten denoting that the thread crashed.*/);
```
Comment (kripken) [inline src/worker.js]: Good point about the wrong stack trace being shown.

How about this: At the top of this `else` body, we know that this isn't an infinite loop etc., and then we want to do some changes like these Atomics and the futex stuff. How about if before the atomics, we check if we managed to set up the thread properly, that is, move your if to up here, and put it behind ASSERTIONS. So something like
```
#if ASSERTIONS
    if (typeof({{{ makeAsmGlobalAccessInPthread('_emscripten_futex_wake') }}}) !== "function") {
      err("could not set up thread properly");
      throw e;
   }
#endif
```

Comment (VirtualTim) [reply src/worker.js]: Done.

**Comment by kripken** [comment]
Great, thanks! Sorry for the amount of time this took, but it's important to get it right.

Odd errors on CI, but I'm sure they can't be because of this, merging.
